### PR TITLE
Fix Subscription Config Inheritance For Event Delivery

### LINKED ISSUE #956: [BUG] Project being created without retention policy
Retention policy and other configs when creating projects aren't required. In the situation where a user doesn't set the retention policy when creating a project, a 400 error is returned when updating the project with a message: `policy:please provide a valid retention policy`

### LINKED ISSUE #974: [BUG] Endpoint rate limit misconfiguration.
Rate limiting to endpoints should use the rate limiting configuration specified in the subscription. Currently, it uses the deprecated endpoint rate limiting configuration.

### PR DESCRIPTION
**Pull request by Dotunj**
- Added a new `RateLimitConfiguration` on the Subscription Model

Resolves #956 #973 #974 #975

### CODE REVIEW

File: server/models/models.go [hunk 1]
```diff
@@ -261,9 +263,18 @@ type UpdateSubscription struct {
...
+	FilterConfig    *datastore.FilterConfiguration    `json:"filter_config,omitempty"`
+	RateLimitConfig *datastore.RateLimitConfiguration `json:"rate_limit_config,omitempty"`
+	DisableEndpoint *bool                             `json:"disable_endpoint" bson:"disable_endpoint"`
+}
+
+type RetryConfiguration struct {
+	Type            datastore.StrategyProvider `json:"type,omitempty" valid:"supported_retry_strategy~please provide a valid retry strategy type"`
+	Duration        string                     `json:"duration,omitempty" valid:"duration~please provide a valid time duration"`
+	IntervalSeconds uint64                     `json:"interval_seconds" valid:"int~please provide a valid interval seconds"`
+	RetryCount      uint64                     `json:"retry_count" valid:"int~please provide a valid retry count"`
```
Comment (jirevwe) [inline server/models/models.go:277]: Why is this being defined here, I thought there it was already defined in the models package?

Comment (Subomi) [reply server/models/models.go:276]: @jirevwe We can resolve this comment yeah?

File: worker/task/process_event_delivery.go [hunk 2]
```diff
@@ -300,3 +280,65 @@ func parseAttemptFromResponse(m *datastore.EventDelivery, e *datastore.Endpoint,
...
+
+func (ec *EventDeliveryConfig) retryConfig() (*RetryConfig, error) {
+	rc := &RetryConfig{}
+
+	if ec.subscription.RetryConfig != nil {
+		rc.Duration = ec.subscription.RetryConfig.Duration
+		rc.RetryCount = ec.subscription.RetryConfig.RetryCount
+		rc.Type = ec.subscription.RetryConfig.Type
+
+	} else if ec.group.Config != nil {
```
Comment (Subomi) [inline worker/task/process_event_delivery.go:316]: @Dotunj Group Config should never be `nil`. At the point of creation, if the user does not set it, we should apply the default to the project. 

It should be an error scenario that a group has no configuration. It's similar to #956

File: worker/task/process_event_delivery.go [hunk 3]
```diff
@@ -300,3 +280,65 @@ func parseAttemptFromResponse(m *datastore.EventDelivery, e *datastore.Endpoint,
...
+	return rc, nil
+}
+
+func (ec *EventDeliveryConfig) rateLimitConfig() *RateLimitConfig {
+	rlc := &RateLimitConfig{}
+
+	if ec.subscription.RateLimitConfig != nil {
+		rlc.Count = ec.subscription.RateLimitConfig.Count
+		rlc.Duration = ec.subscription.RateLimitConfig.Duration
+	} else if ec.group.Config != nil {
```
Comment (Subomi) [inline worker/task/process_event_delivery.go:335]: @Dotunj Same thing here. :)

**Review by Subomi — CHANGES_REQUESTED** [review]

**Review by jirevwe — APPROVED** [review]

**Review by jirevwe — APPROVED** [review]

**Review by danvixent — APPROVED** [review]

**Pull request merged**